[alpha.webkit.UncountedLocalVarsChecker] Allow uncounted object references within trivial statements

rniwa opened the pull request:
This PR makes alpha.webkit.UncountedLocalVarsChecker ignore raw references and pointers to a ref counted type which appears within "trival" statements. To do this, this PR extends TrivialFunctionAnalysis so that it can also analyze "triviality" of statements as well as that of functions Each Visit* function is now augmented with withCachedResult, which is responsible for looking up and updating the cache for each Visit* functions.

As this PR dramatically improves the false positive rate of the checker, it also deletes the code to ignore raw pointers and references within if and for statements.

rniwa committed:
[alpha.webkit.UncountedLocalVarsChecker] Allow uncounted object references within trivial statements

This PR makes alpha.webkit.UncountedLocalVarsChecker ignore raw references and pointers to
a ref counted type which appears within "trival" statements. To do this, this PR extends
TrivialFunctionAnalysis so that it can also analyze "triviality" of statements as well as
that of functions Each Visit* function is now augmented with withCachedResult, which is
responsible for looking up and updating the cache for each Visit* functions.

As this PR dramatically improves the false positive rate of the checker, it also deletes
the code to ignore raw pointers and references within if and for statements.

haoNoQ commented:
Ooo interesting. I want to spend a bit more time thinking whether this has to go into every callback, maybe it can be a set-and-forget thing? (Probably not.)

Also it might be a good idea to cache only statements that may ever get directly queried. (This seems to be exactly necessary and sufficient.) For now they're just the 5 control flow statements right? Maybe add caching to these 5 callbacks and skip the rest? (I suspect it may matter for performance as well, dunno hard to tell though.)

rniwa commented:
> I want to spend a bit more time thinking whether this has to go into every callback, maybe it can be a set-and-forget thing? (Probably not.)

What do you mean by set-and-forget?

haoNoQ commented:
> > I want to spend a bit more time thinking whether this has to go into every callback, maybe it can be a set-and-forget thing? (Probably not.)
> 
> What do you mean by set-and-forget?

I mean like, so that you didn't have to remember to edit every callback in a specific way in order to have caching for every callback. But it might also be a bad idea given that we probably don't actually want every callback to be cached. But maybe there's a way to also make it flexible, dunno.

Yeah I don't immediately see how to do that, your solution is probably perfect. Can you confirm that only 5 callbacks currently need to be cached? If yes, it's probably be great to just edit them manually to add caching when necessary, because now it's a conscious decision.

rniwa committed:
Limit the cache to if, for, while, and compound statements.

haoNoQ commented on clang/lib/StaticAnalyzer/Checkers/WebKit/UncountedLocalVarsChecker.cpp:
```diff
@@ -171,6 +151,24 @@ class UncountedLocalVarsChecker
     if (IsUncountedPtr && *IsUncountedPtr) {
+
+      ASTContext &ctx = V->getASTContext();
+      for (DynTypedNodeList ancestors = ctx.getParents(*V); !ancestors.empty();
+           ancestors = ctx.getParents(*ancestors.begin())) {
```
I don't think this should be a recursive loop. As is, it'll match `x` not only in
```
if (int x = 1) {
}
```
but also in
```
if (1) {
  int x;
}
```
which is probably not what you want(?) So you probably need to check that `V` is the condition variable.

Also, generally speaking, usually we try to avoid `ParentMap`/`getParent()` shenanigans. Parent lookups aren't naturally provided by the AST, they rely on building the "parent map" by scanning the entire AST in the natural order and putting all edges into the map. There are a few very confusing cornercases in this algorithm and a few known crashes in that I honestly don't know how to fix. It also screws algorithmic complexity because now we have an additional backwards traversal; it's very easy to accidentally start walking back and forth indefinitely, or just way too many times. Sometimes it's necessary but most of the time there's a normal top-down approach that gives the answer directly.

In this case I think it's better to redefine eg. `VisitIfStmt()`/`TraverseIfStmt()`to manually traverse the condition variable in a custom manner, and then suppress the automatic traversal (or just traverse the initializer). That's clearly O(1) and doesn't rely on any unreliable technology.

rniwa committed:
Merge function and statement caches.

rniwa commented on clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.h:
```diff
@@ -70,15 +71,27 @@ bool isSingleton(const FunctionDecl *F);
+  using FunctionCacheTy = llvm::DenseMap<const Decl *, bool>;
+  mutable FunctionCacheTy TheFunctionCache{};
+
+  using StatementCacheTy = llvm::DenseMap<const Stmt *, bool>;
+  mutable StatementCacheTy TheStatementCache{};
```
Okay, done that.

rniwa committed:
Intercept Traverse* functions to find trivial statements Instead of traversing the AST context up to find trivial statements.

rniwa committed:
Fix formatting.

rniwa committed:
Fix the debug assertion failure observed on bots.

haoNoQ commented on clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp:
```diff
@@ -253,6 +253,17 @@ class TrivialFunctionAnalysisVisitor
     return true;
   }
 
+  template <typename StmtType, typename CheckFunction>
+  bool WithCachedResult(const StmtType *S, CheckFunction Function) {
```
```suggestion
  template <typename CheckFunction>
  bool WithCachedResult(const Stmt *S, CheckFunction Function) {
```
A bit simpler since we don't really care about the sub-class.

rniwa replied:
Good point. Changed.

haoNoQ commented on clang/lib/StaticAnalyzer/Checkers/WebKit/UncountedLocalVarsChecker.cpp:
```diff
@@ -189,18 +202,16 @@ class UncountedLocalVarsChecker
+                   isRefcountedStringsHack(MaybeGuardian)) &&
+                  isGuardedScopeEmbeddedInGuardianScope(V, MaybeGuardian))
+                return;
+            }
```
Nothing really changed here right? Also coding guidelines actually prefer early returns (https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code).

rniwa replied:
So before this PR, those early exits took place when `MaybeGuardianArgType` was `nullptr` or if `MaybeGuardianArgType` wasn't a `CXXRecordDecl`. This PR changes so that we only return when all these conditions hold true or when it's a `ParmVarDecl`.

rniwa replied:
Specifically, without this change, we'd get the following error:
```
clang/test/Analysis/Checkers/WebKit/uncounted-local-vars.cpp Line 162 (directive at clang/test/Analysis/Checkers/WebKit/uncounted-local-vars.cpp:163):
Local variable 'c' is uncounted and unsafe [alpha.webkit.UncountedLocalVarsChecker]
```

haoNoQ replied:
Oh right ok!

haoNoQ commented on clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp:
```diff
@@ -474,4 +504,22 @@ bool TrivialFunctionAnalysis::isTrivialImpl(
 
+bool TrivialFunctionAnalysis::isTrivialImpl(
+    const Stmt *S, TrivialFunctionAnalysis::CacheTy &Cache) {
+  // If the statement isn't in the cache, conservatively assume that
+  // it's not trivial until analysis completes. Unlike a function case,
+  // we don't insert an entry into the cache until Visit returns
+  // since Visit* functions themselves make use of the cache.
```
>  ```
> // If the statement isn't in the cache, conservatively assume that
> // it's not trivial until analysis completes.
> ```

This comment should probably be moved to `WithCachedResult()` where the false-by-default insertion actually takes place. (Not sure if it was clear that my original comment was talking about that insertion specifically. If it wasn't clear, it was probably a bad comment that needed to be phrased differently in the other function as well.)

> ```
> //                                            Unlike a function case,
> // we don't insert an entry into the cache until Visit returns
> // since Visit* functions themselves make use of the cache.
> ```

We can probably avoid double insert/lookup by skipping this insert/lookup entirely. Just rely on `Visit()` to do the bookkeeping.

But at the same time, it's nice that your code guarantees *some* caching even when we forget to cache a specific statement. It's a property we'd lose if we simply remove this global insert-lookup.

So I wonder if you can replace it with an *assertion* instead? Something like:
```
bool Result = V.Visit(S);
assert(Cache.contains(S) && "Top-level statement not properly cached!");
return Result;
```
This may actively help us find forgotten `WithCachedResult()` calls instead of silently accepting them.

haoNoQ reviewed (commented):
Ooo looks great now!

One more round of nitpicks, but I think it's mostly ready to land.

rniwa committed:
Moved the comment per review comment.

Also made TrivialFunctionAnalysisVisitor's Cache a reference instead of a copy so that
TrivialFunctionAnalysis::isTrivialImpl can retrieve the result out of the cache.

rniwa commented:
Actually, we need to cache the results for `VisitDeclRefExpr` as well to avoid infinite recursion.

rniwa committed:
Also cache the results for VisitDeclRefExpr as a self referencing variable declaration
could end up in an infinite recursion without it.

haoNoQ commented on clang/lib/StaticAnalyzer/Checkers/WebKit/PtrTypesSemantics.cpp:
```diff
@@ -305,19 +337,21 @@ class TrivialFunctionAnalysisVisitor
+          if (VD->hasConstantInitialization() && VD->getEvaluatedValue())
+            return true;
+          auto *Init = VD->getInit();
+          return !Init || Visit(Init);
```
> Actually, we need to cache the results for `VisitDeclRefExpr` as well to avoid infinite recursion.

Wait, hmmm, this looks unusual. Why are we visiting the initializer whenever the variable is *accessed*? Isn't it sufficient to visit the initializer whenever the object is *initialized*? The initializer isn't evaluated at use site and it doesn't necessarily represent the contents of the variable anymore. Why do we care what the variable contained "initially"?

It's good that we aren't doing it *every* time the variable is used (so we can treat it as an unrelated problem for another patch) but we probably weren't doing it at the right time in the first place.

rniwa replied:
Oh, that is a very good point. I guess we can just always `return true` for `DeclRefExpr`.

haoNoQ reviewed (approved):
LGTM!

I think we have to look at the `DeclRefExpr` case more closely - something seems fishy about it, but also this isn't this patch's fault.

rniwa committed:
Make VisitDeclRefExpr always return true instead of checking the initializer's triviality.
